### PR TITLE
Improve Ball timestamping and prediction

### LINKED ISSUE #49: Improve timestamping system
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
Our current timestamping system is not very well fleshed out and I think could be improved significantly. In addition we should be using it to pass around times/durations to functions (such as for predicting future ball positions) since it's more robust than using ints.

**Describe the solution you'd like**
* I think `AITimestamp` should be a `std::chrono::time_point` instead of `duration`
* Then we could have a separate `duration`. The separation makes things more clear
* We could probably use a few utility functions for common functions
  * turning durations into seconds / milliseconds
  * easy creating of durations
  * creation of durations from timestamps

This should all be using the `std::chrono` library as the backend

### PR DESCRIPTION
**Pull request by MathewMacDougall**
Update the Ball class to use `std::chrono` for timestamps. Also added a new `updateState` function to update the Ball's state to a predicted future state. Useful for when the AI node runs several times without getting updates from vision, but we want to act with the most up-to-date/accurate information.

If anyone has naming suggestions to make the different `update` functions more clear I'm all ears.

I'm going to roll out a few timestamp changes before finishing the last `Team` and `World` implementations. This is the first part.

partially resolved #49

### CODE REVIEW

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looks pretty good, I like the design decisions made with respect to timestamps here.

File: src/thunderbots/software/ai/world/ball.h [hunk 1]
```diff
@@ -31,6 +39,21 @@ class Ball final
      */
     void update(const Ball& new_ball_data);
 
+    /**
+     * Updates the ball's state to match what it would be at the given timestamp. The
+     * timestamp must be >= the ball's last update timestamp.
+     *
+     * @param timestamp The timestamp at which to update the ball's state to
+     */
+    void updateState(std::chrono::time_point<std::chrono::steady_clock> timestamp);
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/world/ball.h]: Yes makes sense. I like that wording better. I think I'll update the other update functions to `udpateState` too just to be explicit

File: src/thunderbots/software/test/world/ball.cpp [hunk 2]
```diff
@@ -1,56 +1,121 @@
...
+    // Can't compare timestamp values here because the ball and the expected timestamp
+    // would not be createdx at the same time, and would not be equal. We could check
+    // that the timestamps are within a certain threshold, but that is not robust and
+    // makes the test dependant on the speed of the system executing it
 }
 
 TEST(BallTest, construct_with_params)
 {
-    Ball ball = Ball(Point(1, 2.3), Vector(-0.04, 0.0));
+    auto current_time = steady_clock::now();
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/thunderbots/software/test/world/ball.cpp]: Looks like you sort of can by creating time points relative to the epoch.
https://stackoverflow.com/questions/24913749/c11-chrono-create-time-point-from-number
https://stackoverflow.com/questions/30756919/convert-chrono-duration-to-time-point

Think it's worth?

[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/test/world/ball.cpp]: Yeah, you can set up a time once, like:
```
class BallTest : public ::testing::Test {
 protected:
    // You can define per-test set-up logic as usual.
  virtual void SetUp() { ... }

  // You can define per-test tear-down logic as usual.
  virtual void TearDown() { ... }

  // Some expensive resource shared by all tests. (TIME WOULD GO HERE)
  static T* shared_resource_;
};
```
and then making the tests look like:
```
TEST_F(BallTest, construct_with_params){
    shared_resource_.blahBlah()
    ....
}
```

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**